### PR TITLE
fix: clone api doesn't take update configurations

### PR DESCRIPTION
**Pull request by harshithmullapudi**
## What
Clone API now can take the configuration and override it with the existing configuration

https://github.com/airbytehq/airbyte-internal-issues/issues/366

### CODE REVIEW

File: airbyte-api/src/main/openapi/config.yaml [hunk 1]
```diff
@@ -2692,6 +2692,24 @@ components:
...
+        sourceCloneId:
+          $ref: "#/components/schemas/SourceId"
+        sourceConfiguration:
+          $ref: "#/components/schemas/SourceCloneConfiguration"
+    SourceCloneConfiguration:
+      type: object
+      properties:
+        connectionConfiguration:
+          $ref: "#/components/schemas/SourceConfiguration"
+        workspaceId:
```
Comment (cgardens) [inline airbyte-api/src/main/openapi/config.yaml:2709]: why do we want to let them update workspace id? won't that potentially give other workspaces access to other connectors?

**Review by cgardens — CHANGES_REQUESTED** [review]

File: airbyte-api/src/main/openapi/config.yaml (hunk 1 above)
Comment (harshithmullapudi) [reply airbyte-api/src/main/openapi/config.yaml:2709]: That's true. I thought what if we want to clone a source into other workspace?

But yeah we can remove that feature for now 😅

**Comment by timroes** [comment]
I'm setup as a reviewer here, but I lack some context? What issue is this solving or what's exactly the purpose of this? Do we have an open issue to reference here?

**Comment by harshithmullapudi** [comment]
@timroes I have updated the PR with the issue link

File: airbyte-api/src/main/openapi/config.yaml [hunk 2]
```diff
@@ -2692,6 +2692,22 @@ components:
       properties:
         sourceId:
           $ref: "#/components/schemas/SourceId"
+    SourceCloneRequestBody:
```
Comment (cgardens) [inline airbyte-api/src/main/openapi/config.yaml:2695]: can you add a description field for `SourceCloneRequestBody` and `DestinationCloneRequestBody`?

**Review by cgardens — APPROVED** [review]
Have you tested it out manually to verify it works?

**Comment by harshithmullapudi** [comment]
> Have you tested it out manually to verify it works?

Yeah I verified by building and running the server,

File: airbyte-api/src/main/openapi/config.yaml (hunk 2 above)
Comment (harshithmullapudi) [reply airbyte-api/src/main/openapi/config.yaml:2695]: Will do.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/DestinationHandler.java [hunk 3]
```diff
@@ -168,26 +170,32 @@ public DestinationRead getDestination(final DestinationIdRequestBody destination
     return buildDestinationRead(destinationIdRequestBody.getDestinationId());
   }
 
-  public DestinationRead cloneDestination(final DestinationIdRequestBody destinationIdRequestBody)
+  public DestinationRead cloneDestination(final DestinationCloneRequestBody destinationCloneRequestBody)
       throws JsonValidationException, IOException, ConfigNotFoundException {
     // read destination configuration from db
-    final DestinationRead destinationToClone = buildDestinationReadWithSecrets(destinationIdRequestBody.getDestinationId());
-    final ConnectorSpecification spec = getSpec(destinationToClone.getDestinationDefinitionId());
+    final DestinationRead destinationToClone = buildDestinationReadWithSecrets(destinationCloneRequestBody.getDestinationCloneId());
```
Comment (cgardens) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/DestinationHandler.java:176]: we should verify that the source / destination being cloned is from the same workspace.

**Review by cgardens — CHANGES_REQUESTED** [review]
we should be scoping these clones by workspace otherwise it won't be safe to expose in cloud. not a problem yet, since this functionality is not expose in cloud but it needs to be fixed before it is.

**Review by cgardens — APPROVED** [review]
Discussed offline and agreed that the current implementation is safe. When ported to cloud we need to make sure that the user has access to the source / destination being cloned, which should be pretty obvious / straightforward when we get there.

**Pull request merged**